From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Sync attack strength

Split the attack strength ticking out of Player#tick so it can be
synchronised across servers, and broadcast a reset of the attack
strength ticker to the other servers.

diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 8a65f72..0f4563e 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -123,6 +123,7 @@ import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Team;
 import org.jspecify.annotations.Nullable;
+import puregero.multipaper.externalserverprotocol.PlayerResetAttackStrengthPacket;
 
 public abstract class Player extends Avatar implements ContainerUser {
     public static final int MAX_HEALTH = 20;
@@ -314,7 +315,17 @@ public abstract class Player extends Avatar implements ContainerUser {
             this.setPos(nx, this.getY(), nz);
         }
 
+        // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         this.attackStrengthTicker++;
+    }
+
+    public void tickAttackStrengthAfter() {
+        // MultiPaper end
         this.itemSwapTicker++;
         ItemStack mainHandItemStack = this.getMainHandItem();
         if (!ItemStack.matches(this.lastItemInMainHand, mainHandItemStack)) {
@@ -2033,6 +2045,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     public void resetAttackStrengthTicker() {
+        PlayerResetAttackStrengthPacket.broadcastResetAttackStrength(this);
         this.attackStrengthTicker = 0;
         this.itemSwapTicker = 0;
     }
